Fix esm

davidlehn opened the pull request.

davidlehn committed:
Fix examples.

davidlehn committed:
Switch from Travis CI to GitHub Actions.

davidlehn committed:
Add default exports.

- Helps with some import patterns in other module and commonjs code.

davidlehn committed:
Updates ESM usage.

- **BREAKING**: Update Node.js requirement to v10. Due to modern tools
  dropping older Node.js version support. The code itself may still work
  with older Node.js versions.
- Set package `type` to `module`.
- Add package `exports` for `import` and `require`.
- Change `.js` to `.cjs` as needed.
- Update dependencies.
- Switch test environment detection to node/browser specific files.
- Switch tests to ESM.

davidlehn committed:
Drop Node.js v10 support.

- Fixes ESM testing problems in v10. Code likely still works with
  earlier versions and the 'esm' package.

dmitrizagidulin commented on tests/api.spec.js:
```diff
@@ -1,17 +1,13 @@
+import chaiBytes from 'chai-bytes';
+chai.use(chaiBytes);
 const should = chai.should();
+import {TextEncoder} from './util.js';
```
I think we can just drop this, since `TextEncoder` & decoder is definitely available in Node.js global in Node 12+

davidlehn replied:
On second thought... I'll just remove that.  It's just test code and we can't test with <v12 now anyway.  It was just there thinking v10 testing still worked.

davidlehn commented on main.js:
```diff
@@ -18,3 +16,8 @@ export function encode(input, maxline) {
   return _decode(input, alphabet);
 }
+
+export default {
```
It was added due to patterns like this:
```js
import base58btc from 'base58-universal';
```
(Which is a coming-soon patch to the current `require` code here:
https://github.com/digitalbazaar/ed25519-signature-2020/blob/57396fc1384ba848f93b3da27c8069b132fb7790/lib/Ed25519Signature2020.js#L6)

We could convert that to importing the functions directly, but if people want to pull it all in at once, I think we need the `default` exports.  I'm not sure what best practices are for this.

dmitrizagidulin replied:
Ah, I see. Well, I'd vote for removing the default export (& releasing this as a breaking change)

davidlehn committed:
Use global TextEncoder.

Tests now only run in Node.js v12+ so the global can be used as in
browsers.

davidlehn committed:
Convert to module (ESM).

- **BREAKING**: Convert to module (ESM).
- **BREAKING**: Require Node.js >=14.
- Update dependencies.
- Lint module.

davidlehn committed:
Update node versions.

davidlehn committed:
Remove karma babel processor.

davidlehn committed:
Use files globstar.

davidlehn committed:
Update dependencies.

davidlehn committed:
Update coverage commands.

davidlehn committed:
Update README.
